File: crates/desktop/frontend/src/chat-context.ts
/**
 * Pure helpers for "Add context to chat" (phase 17): turn what the user is
 * looking at — a terminal selection, the active file, its git diff, an
 * editor selection — into a fenced block the chat composer can hold, and
 * parse tool messages back into card data. No DOM, no IPC; vitest-covered.
 */

export type ContextKind = "terminal" | "file" | "diff" | "editor-selection";

/** Max lines a single injected block may carry before it is cut. */
export const CONTEXT_MAX_LINES = 200;

export const TRUNCATED_MARKER = "…truncated";

export interface ContextMeta {
  /** Tab label (terminal) or workspace-relative path (file / diff / selection). */
  name: string;
  /** 1-based inclusive line range, when only part of a file is included. */
  lines?: { from: number; to: number };
}

/** Header line of a block, e.g. `File: src/main.rs (lines 10–42)`. */
export function contextHeader(kind: ContextKind, meta: ContextMeta): string {
  const range = meta.lines ? ` (lines ${meta.lines.from}–${meta.lines.to})` : "";
  switch (kind) {
    case "terminal":
      return `Terminal selection (tab "${meta.name}")`;
    case "file":
      return `File: ${meta.name}${range}`;
    case "diff":
      return `Diff: ${meta.name}`;
    case "editor-selection":
      return `Selected text in editor: ${meta.name}${range}`;
  }
}

/** Cut `text` to `max` lines, appending the truncation marker when it did. */
export function truncateLines(text: string, max = CONTEXT_MAX_LINES): { text: string; truncated: boolean } {
  const lines = text.replace(/\r\n?/g, "\n").split("\n");
  if (lines.length <= max) return { text, truncated: false };
  return { text: lines.slice(0, max).join("\n") + "\n" + TRUNCATED_MARKER, truncated: true };
}

/** Fence language hint from a path's extension ("" when unknown). */
export function fenceLang(kind: ContextKind, name: string): string {
  if (kind === "diff") return "diff";
  if (kind === "terminal") return "text";
  const ext = name.split(".").pop()?.toLowerCase() ?? "";
  return /^[a-z0-9]{1,10}$/.test(ext) && ext !== name.toLowerCase() ? ext : "";
}

/** A fence that cannot be closed early by the content: one more backtick
 *  than the longest run inside `text` (min 3). */
export function fenceFor(text: string): string {
  let longest = 0;
  for (const m of text.matchAll(/`+/g)) longest = Math.max(longest, m[0].length);
  return "`".repeat(Math.max(3, longest + 1));
}

/**
 * The block the composer receives: header line, then a fenced (possibly
 * truncated) body. Trailing whitespace of the body is dropped so the fence
 * closes tight; an empty body yields an empty string (nothing to inject).
 */
export function fenceBlock(kind: ContextKind, meta: ContextMeta, text: string): string {
  const body = text.replace(/\s+$/, "");
  if (body.length === 0) return "";
  const cut = truncateLines(body);
  const fence = fenceFor(cut.text);
  return `${contextHeader(kind, meta)}\n${fence}${fenceLang(kind, meta.name)}\n${cut.text}\n${fence}\n`;
}

/** Join a block onto the composer's current draft with a blank line between. */
export function appendToDraft(draft: string, block: string): string {
  if (!block) return draft;
  const head = draft.replace(/\s+$/, "");
  return head.length === 0 ? block : `${head}\n\n${block}`;
}

/** Flatten the backend's `DiffLine[]` (get_file_diff) to unified-diff text. */
export function diffLinesToText(lines: { content: string; line_type: string }[]): string {
  return lines
    .map((l) => {
      switch (l.line_type) {
        case "add":
          return `+${l.content}`;
        case "del":
          return `-${l.content}`;
        case "context":
          return ` ${l.content}`;
        default:
          return l.content;
      }
    })
    .join("\n");
}

// ── Tool cards ─────────────────────────────────────

export type ToolCardStatus = "running" | "ok" | "error" | "approval" | "approved" | "denied";

export interface ToolCard {
  id: string;
  name: string;
  /** Pretty-printed JSON arguments ("" when unknown). */
  args: string;
  result: string;
  status: ToolCardStatus;
  durationMs?: number;
}

/** Tool result content the backend stores in history: `[name] [Error] text`.
 *  Parsed back so a reloaded conversation still renders cards. */
export function parseToolMessage(content: string): { name: string; result: string; isError: boolean } {
  const m = /^\[([^\]\s]+)\] (\[Error\] )?([\s\S]*)$/.exec(content);
  if (!m) return { name: "tool", result: content, isError: false };
  return { name: m[1], result: m[3], isError: m[2] !== undefined };
}

/** Pretty JSON for a card body; unparsable input is returned as-is. */
export function prettyJson(value: unknown): string {
  if (typeof value === "string") {
    try {
      return JSON.stringify(JSON.parse(value), null, 2);
    } catch {
      return value;
    }
  }
  try {
    return JSON.stringify(value, null, 2);
  } catch {
    return String(value);
  }
}

/** Results longer than this many lines start collapsed ("Show more"). */
export const RESULT_COLLAPSE_LINES = 12;

export function formatDurationMs(ms: number): string {
  if (ms < 1000) return `${Math.round(ms)} ms`;
  if (ms < 60_000) return `${(ms / 1000).toFixed(1)} s`;
  return `${Math.floor(ms / 60_000)} m ${Math.round((ms % 60_000) / 1000)} s`;
}

/** The chat-context chooser: which entries are enabled for the current view. */
export interface ContextAvailability {
  terminalSelection: boolean;
  activeFile: boolean;
  editorSelection: boolean;
}

export function contextChoices(a: ContextAvailability): { kind: ContextKind; label: string; disabled: boolean }[] {
  return [
    { kind: "terminal", label: "Terminal selection", disabled: !a.terminalSelection },
    { kind: "file", label: "Active file", disabled: !a.activeFile },
    { kind: "diff", label: "Diff of active file", disabled: !a.activeFile },
    { kind: "editor-selection", label: "Selected text in editor", disabled: !a.editorSelection },
  ];
}

File: crates/tui/src/chat_context.rs
//! Turning what the user is looking at — a terminal selection, a file — into
//! a fenced block the chat composer can hold. The TUI half of the desktop's
//! `chat-context.ts`: same header lines, same 200-line cap, same
//! can't-break-out fence, so a conversation reads identically whichever
//! frontend fed it. Pure — no `App`, no IO.

/// Max lines a single injected block may carry before it is cut.
pub(crate) const CONTEXT_MAX_LINES: usize = 200;

pub(crate) const TRUNCATED_MARKER: &str = "…truncated";

/// What the block holds; picks the header wording and the fence hint.
pub(crate) enum ContextKind {
    /// A selection in a terminal tab; `name` is the tab label.
    Terminal,
    /// A file; `name` is its workspace-relative path.
    File,
}

/// Header line of a block, e.g. `File: src/main.rs`.
pub(crate) fn context_header(kind: &ContextKind, name: &str) -> String {
    match kind {
        ContextKind::Terminal => format!("Terminal selection (tab \"{name}\")"),
        ContextKind::File => format!("File: {name}"),
    }
}

/// Cut `text` to `max` lines, appending the truncation marker when it did.
pub(crate) fn truncate_lines(text: &str, max: usize) -> String {
    let normalized = text.replace("\r\n", "\n").replace('\r', "\n");
    let lines: Vec<&str> = normalized.split('\n').collect();
    if lines.len() <= max {
        return normalized;
    }
    format!("{}\n{TRUNCATED_MARKER}", lines[..max].join("\n"))
}

/// Fence language hint from a path's extension (empty when unknown).
fn fence_lang(kind: &ContextKind, name: &str) -> String {
    match kind {
        ContextKind::Terminal => "text".to_string(),
        ContextKind::File => match name.rsplit_once('.') {
            Some((stem, ext))
                if !stem.is_empty()
                    && (1..=10).contains(&ext.len())
                    && ext.chars().all(|c| c.is_ascii_alphanumeric()) =>
            {
                ext.to_ascii_lowercase()
            }
            _ => String::new(),
        },
    }
}

/// A fence that the content cannot close early: one more backtick than the
/// longest run inside `text` (minimum 3).
fn fence_for(text: &str) -> String {
    let mut longest = 0usize;
    let mut run = 0usize;
    for c in text.chars() {
        if c == '`' {
            run += 1;
            longest = longest.max(run);
        } else {
            run = 0;
        }
    }
    "`".repeat(longest.saturating_add(1).max(3))
}

/// The block the composer receives: header line, then a fenced (possibly
/// truncated) body. Trailing whitespace is dropped so the fence closes tight;
/// an empty body yields an empty string — there is nothing to inject.
pub(crate) fn fence_block(kind: &ContextKind, name: &str, text: &str) -> String {
    let body = text.trim_end();
    if body.is_empty() {
        return String::new();
    }
    let cut = truncate_lines(body, CONTEXT_MAX_LINES);
    let fence = fence_for(&cut);
    format!(
        "{}\n{fence}{}\n{cut}\n{fence}\n",
        context_header(kind, name),
        fence_lang(kind, name),
    )
}

/// Join a block onto the composer's current draft with a blank line between.
pub(crate) fn append_to_draft(draft: &str, block: &str) -> String {
    if block.is_empty() {
        return draft.to_string();
    }
    let head = draft.trim_end();
    if head.is_empty() {
        block.to_string()
    } else {
        format!("{head}\n\n{block}")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn header_wording_matches_the_desktop() {
        assert_eq!(
            context_header(&ContextKind::Terminal, "Claude Code"),
            "Terminal selection (tab \"Claude Code\")"
        );
        assert_eq!(
            context_header(&ContextKind::File, "src/main.rs"),
            "File: src/main.rs"
        );
    }

    #[test]
    fn a_short_body_is_kept_whole() {
        let out = truncate_lines("a\nb\nc", 200);
        assert_eq!(out, "a\nb\nc");
    }

    #[test]
    fn a_long_body_is_cut_and_marked() {
        let text = (0..250)
            .map(|i| i.to_string())
            .collect::<Vec<_>>()
            .join("\n");
        let out = truncate_lines(&text, 200);
        let lines: Vec<&str> = out.split('\n').collect();
        assert_eq!(lines.len(), 201, "200 lines plus the marker");
        assert_eq!(lines[199], "199");
        assert_eq!(lines[200], TRUNCATED_MARKER);
    }

    #[test]
    fn crlf_is_normalized_before_counting() {
        assert_eq!(truncate_lines("a\r\nb", 200), "a\nb");
    }

    /// A body containing a fence must not be able to end the block early.
    #[test]
    fn the_fence_outgrows_the_content() {
        let block = fence_block(&ContextKind::Terminal, "sh", "```\ncode\n```");
        assert!(
            block.contains("````"),
            "expected a 4-backtick fence, got:\n{block}"
        );
    }

    #[test]
    fn an_empty_body_injects_nothing() {
        assert_eq!(fence_block(&ContextKind::Terminal, "sh", "   \n\n"), "");
        assert_eq!(append_to_draft("hi", ""), "hi");
    }

    #[test]
    fn a_file_block_carries_its_language_hint() {
        let block = fence_block(&ContextKind::File, "src/main.rs", "fn main() {}");
        assert!(block.starts_with("File: src/main.rs\n```rs\n"), "{block}");
        assert!(block.ends_with("```\n"), "{block}");
        // An extension-less file gets no hint.
        let plain = fence_block(&ContextKind::File, "Makefile", "all:");
        assert!(plain.contains("\n```\nall:"), "{plain}");
    }

    #[test]
    fn a_block_lands_after_what_was_already_typed() {
        let draft = append_to_draft("look at this", "File: a.rs\n```rs\nx\n```\n");
        assert_eq!(draft, "look at this\n\nFile: a.rs\n```rs\nx\n```\n");
        // Onto an empty composer it is the whole draft.
        assert_eq!(append_to_draft("  ", "block\n"), "block\n");
    }
}
